### DISCUSSION TITLE
Collect requirements/idea/features for v2.7.0

### DISCUSSION THREAD

**Discussion by OrlinVasilev** [goharbor/harbor#16775] [General]
Hello Community,

as we released  v2.5 and we have plan and roadmap for v2.6 it's time to brainstorm the ideas and requirements for the feature list for 2.7!

Please share your thoughts here so we can all work on proper roadmap and achieving it!
Keep in mind that Harbor is community developed and driven project so not everything can be implemented right away :) 
If you have an idea or need and eager to work on, we will support you and help you get there!

Thank you being essential part of the community!

Orlix
Harbor Community Manager

**Comment by AlexBarth13** [comment]
We would appreciate a working API concept where "normal" users and robot accounts/technical users can make automated API calls to Harbor

> Edit: added _and robot accounts/technical users_

<br ><br >
Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Mercedes-Benz Tech Innovation GmbH, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by AlexBarth13** [comment]
Enable that e.g. projects/repositories can be opened in a new browser tab
<br ><br >
Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Mercedes-Benz Tech Innovation GmbH, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by ekarlso** [comment]
Would it be poasible to add user federation concept to harbor where you could use something like girhub federated identity to login?

**Reply by shawngmc** [reply]
Would it make sense to integrate that directly into Harbor or use Keycloak as an OIDC provider for Harbor, and let Keycloak handle the bridging to external ID providers (Google, GitHub, etc.)?

**Comment by shawngmc** [comment]
A few for now, might have some more ideas later:

- Artifact Vulnerability UI: Button to download the existing JSON report (as from the /projects/{project_name}/repositories/{repository_name}/artifacts/{reference}/additions/vulnerabilities API endpoint)
- Artifact Vulnerability UI: Some sort of HTML export option. The current UI is fine for quick glances, but a nice report to share/comment on/not need to expand every section would be nice.
- More explicit options to configure Harbor from the YML file; for example, the OIDC connection has to be supplied in the UI and I have issues where the client secret is 'lost' when redeploying because the encryption secret has changed.
- Improved documentation of the contents after the prepare/redeploying with them

**Reply by AllForNothing** [reply]
CVE data exporting feature is on process https://github.com/goharbor/community/pull/174

**Comment by AlexBarth13** [comment]
Just for convenient reasons:
- Allow to use the links in the documentation without the version number in it and automatically be redirected to the latest version.

Example:
Working: https://goharbor.io/docs/2.4.0/administration/configure-proxy-cache/
Not working: https://goharbor.io/docs/administration/configure-proxy-cache/

Example from Tanzu:
https://docs.pivotal.io/application-service/concepts/overview.html will be automatically redirected to
https://docs.pivotal.io/application-service/2-12/concepts/overview.html
<br ><br >
Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Mercedes-Benz Tech Innovation GmbH, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Reply by OrlinVasilev** [reply]
I was thinking the same in the past, and that is not related to the release anyway :)
Can you create an issue in the website repo or should I?

**Reply by AlexBarth13** [reply]
Thanks for pointing this out!
Of course you are totally correct, that this is not related to an release - sorry!
I will create an GitHub Issue for this :)
<br ><br >
Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Mercedes-Benz Tech Innovation GmbH, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Reply by zyyw** [reply]
It seems that https://goharbor.io/docs will be automatically redirected to the latest version of doc, for example https://goharbor.io/docs/2.5.0, while subpages will not be redirected, for example https://goharbor.io/docs/administration/configure-proxy-cache/ will not be redirected and report 404.

**Comment by thangamani-arun** [comment]
Hi,

I know harbor has geo-replication feature, but it is currently replicating only the repositories. whatever goes to database like users, projects. quota's etc are not replicated. We use postgres streaming replication to remote site, but the local harbor-core instances requires a WRITABLE DB on a read-only postgres replica pod, which is not possible unless it got promoted to master mode.

It possible to have active-standby across DC-DR sites, so everything gets replicated across sites. In this case, we don't have to rely on DB replication. DB can be updated using harbor replication feature beyond registry replication?

**Comment by Vad1mo** [comment]
I would like to see some improvements regarding the API. When you work with the API a lot of things only make sense in the frontend's context and hence a designed like this. 

There are quite a few issues around already. I plan to create a summarizing issue or start a discussion.

**Reply by qnetter** [reply]
Please so. The API should serve the UI, but design should be API-first.

**Comment by AlexBarth13** [comment]
Working webhook integration between Harbor and Mattermost. Currently, Harbor is not able to send a message to Mattermost because the format of the message is wrong (for Mattermost).
Details can be found here: https://github.com/goharbor/harbor/issues/16632#issuecomment-1128450201
<br ><br >
Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Mercedes-Benz Tech Innovation GmbH, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by Vad1mo** [comment]
From the UX/DX point of view and considering the bigger role of OCI artifacts in the Cloud Native ecosystem. 
We should extend Harbor to display different typed of OCI artifact in different groups (tabs)

There are ton of projects already using OCI registries to store other artifacts then container images, eg. ORAS, Helm OCI, Singularity, ML Data Models and so on. 

At the Moment we only have two groups/tabs 
Repositories and Helm Charts. The Helm Charts stored in Chartmuseum are displayed in their own group/tab
<img width="1853" alt="Bildschirmfoto 2022-05-28 um 12 15 48" src="https://user-images.githubusercontent.com/1492007/170821152-13146560-4065-4c18-9d72-041e924d79d9.png">


My proposal for Harbor 2.7 is that we should extend Harbor in a way that the user can decide on a project basis what groups he wants to see. There should be a default list in the project like **Containers**, **Helm Chart** and **OCI Artifacts**  but the user should be able to change that according to the image metadata. 

<img width="1773" alt="Bildschirmfoto 2022-05-28 um 12 18 48" src="https://user-images.githubusercontent.com/1492007/170821142-a074a0dd-1e69-4687-bc00-b4ea377fe8ae.png">


The results would be that an ML team could config their project to only see ML Data to TF and the software  engineering team would be fine with the defaults.

**Comment by AlexBarth13** [comment]
I would like to have the option to add a banner where I can add some information (e.g. maintenance information,...). It could look like this:

![image](https://user-images.githubusercontent.com/99802718/175266721-e1ce0c63-194c-4785-9d9d-b68577a90358.png)


<br ><br >
Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Mercedes-Benz Tech Innovation GmbH, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Reply by shawngmc** [reply]
This would be a big help for managing an instance for multiple client teams. I'd love to take it a step further. Add:

- A customizable text box (maybe the ability to add links) to the login page in the large 'blank' area.
- A persistent banner
- An 'alert' banner that could be dismissed and hidden (until next login or the message changes - I could make arugments for both)

**Comment by jayolmos** [comment]
Hello, it would be nice to have Trivy metrics about CVE in artifacts to integrate them with prometheus + grafana.

**Reply by shawngmc** [reply]
I'd hope that this wouldn't be Trivy-specific, but something gathered through the Harbor Scanner API. My install uses Anchore and the Harbor Anchore Scanner Adapter.

**Comment by OrlinVasilev** [comment]
this one looks must have https://github.com/goharbor/harbor/issues/17125 @dkulchinsky

**Reply by dkulchinsky** [reply]
Thank you @OrlinVasilev, would very much to like to see this being implemented.

This becomes especially important for scaling Harbor and still being able to efficiently scan for vulnerabilities stored images.

I think this may indirectly depend on https://github.com/goharbor/harbor/issues/14638 getting fixed, which seems like there's some possible fix inflight 🤞🏼 🙏🏼

**Reply by slushysnowman** [reply]
Yep this looks like a good call.

**Reply by dkulchinsky** [reply]
looks like this won't make into 2.7 afterall https://github.com/goharbor/harbor/issues/17125#issuecomment-1244732441 😞 but hopefully still on track for 2.8 🙏🏼

**Comment by PengWuLin** [comment]
### The image is too big now
With the upgrade of the go version and the addition of functions, the image is also getting larger。
I want to be able to optimize the image size。
| image       | v2.3.5 | v2.5.3 |
| ----------- | ------ | ------ |
| Core        | 193MB  | 203MB  |
| JobService  | 211MB  | 227MB  |
| Portal      | 58.9MB | 52.6MB |
| Registry    | 82.6MB | 77.9MB |
| Registryctl | 133MB  | 136MB  |
| Trivy       | 133MB  | 251MB  |

**Reply by PengWuLin** [reply]
Is it possible to use the following means

1. Remove debugging information
2. Use UPX tool, UPX is an advanced executable file compressor. UPX will typically reduce the file size of programs and DLLs by around 50%-70%, thus reducing disk space, network load times, download times and other distribution and storage costs.
`
go build -ldflags="-s -w" -o server main.go && upx -9 server
`

**Reply by Vad1mo** [reply]
UPX have quite a few [disadvantages](https://stackoverflow.com/questions/353634/are-there-any-downsides-to-using-upx-to-compress-a-windows-executable), I would rather shrink the package and image size. 

Here are some other methods reduce the size: https://github.com/xaionaro/documentation/blob/master/golang/reduce-binary-size.md

**Reply by PengWuLin** [reply]
thanks

**Comment by shawngmc** [comment]
Configurability/standardization of password requirements

Currently:

- Any account password/token has to be at least 8 characters and contain uppercase, lowercase and numbers
- User generated passwords are only 8 characters long, while manual entry allows 8-30 characters
- Robot generated passwords are 32 characters long, while manual entry allows 8-20 characters
- There are no admin options that modify these settings

For the sake of consistency and security, can we please:

- Add an admin setting for minimum password length enforced across both users and robots
- Make generation consistent between users and robots, following the admin-configured minimum length
- Add an admin setting to allow both, allow only manual entry, or allow only generated tokens

**Comment by OrlinVasilev** [comment]
@sfgets please add your label rework thoughts here :) Thanks !

**Comment by akramelnouby** [comment]
With docker not being supported on RHEL 8+ hosts, I think it would be nice to be able to setup harbor on podman (manually or by using ansible). I know that it shouldn't be that hard, because I am currently working on an ansible playbook and it works (only notary is having issues). It would be nice to see it as an official way to install harbor on nodes running podman as container management. Please also see my Q&A regarding this topic (https://github.com/goharbor/harbor/discussions/17268#discussion-4261406)

**Reply by shawngmc** [reply]
I added a couple questions/ideas to #17268; please take a look and see if these solutions might work.

**Comment by danielzhanghl** [comment]
logging format requirement:


right now Harbor use it's own log library, src/lib/log/*

could consider to use open source logging library like "github.com/Sirupsen/logrus"

or, as Harbor use beego, could use its log as well:
https://github.com/beego/beedoc/blob/master/en-US/module/logs.md


then it will be easy to provide more log options, like json format, out put to file/socket endpoint, like 

https://docs.docker.com/registry/configuration/#log
log:
  accesslog:
    disabled: true
  level: debug
  formatter: text
  fields:
    service: registry
    environment: staging

**Comment by slushysnowman** [comment]
I'd REALLY like to see Harbor admins be able to disable the ability to set projects 'public' on a per-project level (see https://github.com/goharbor/harbor/issues/12306)

The really nice thing about Harbor is that we are able to offer stuff in a self-service way to our teams by giving out project admin rights to certain subgroups of users. But the fact that they can accidentally or otherwise click a button and set their project to be fully public, no auth required to pull images, is super scary.

We'd really like the ability to turn it off, and I think the lack of ability to be able to do so is probably a blocker for some organisations looking to adopt Harbor

**Reply by slushysnowman** [reply]
the 'project' level switch is the minimum I'd like to be able to disable, all the other suggestions in https://github.com/goharbor/harbor/issues/12306 are also very relevant

**Comment by slushysnowman** [comment]
This is also an issue we'd really like to see fixed - https://github.com/goharbor/harbor/issues/14638

We do daily scans with Trivy, but this resets 'last pull' so retention policies based on time since last pull are just broken, meaning we can't do automatic removal of old / unused images. It's a pretty old issue as well, so it'd be great to see it get picked up and fixed.

**Reply by dkulchinsky** [reply]
big +1 on that and I think it would be critical to enable https://github.com/goharbor/harbor/discussions/16775#discussioncomment-3146504 (as I mentioned there as well)

**Comment by AlenversFr** [comment]
Cleaning the technical tables in postgres or more interesting add some harbor parameters to schedule cleaning actions (cron + how many we keep)

linked here [https://github.com/goharbor/harbor/issues/17538](url)

**Comment by dkulchinsky** [comment]
I think gaps in GC have to be addressed

there's a story from last year https://github.com/goharbor/harbor/issues/16010 listing the main issues, some low hanging fruit items were addressed.

However from our experience, GC becomes mostly useless in large scale deployments of Harbor.

Specifically, with the following issues that need to be addressed:

1. https://github.com/goharbor/harbor/issues/12948
2. https://github.com/goharbor/harbor/issues/15807
3. https://github.com/goharbor/harbor/issues/15970

some of these issues are 2+ years old, especially #12948 which in our case results in GC being able to clean only ~200 artifacts in 24 hours, while we have >170,000 artifacts in the thrashed queue.

**Reply by Vad1mo** [reply]
we still see problems popping up with GC.

**Comment by lengrongfu** [comment]
Optimize the push image processing logic when the quota is full. https://github.com/goharbor/harbor/issues/17390
